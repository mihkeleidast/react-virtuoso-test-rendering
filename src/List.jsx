import * as React from 'react';
import {
  Virtuoso
} from 'react-virtuoso';

/**
 * Automatically sized, virtualized vertical infinite list that supports dynamic row height
 */
export const List = React.forwardRef(
  (
    {
      data,
      itemContent,
      itemKey,
      fixedItemHeight,
      estimatedItemHeight,
      overscan,
      className,
      components: propComponents,
      initialItemCount,
      useWindowScroll,
      scrollParent,
      context,
      stickyItemCount,
      ...rest
    },
    ref,
  ) => {
    const renderItemContent = React.useCallback(
      (index, item, context) => {
        return React.createElement(itemContent, { item, index, context });
      },
      [itemContent],
    );

    const components = React.useMemo(
      () => ({
        List: propComponents?.List,
        Item: propComponents?.Item,
      }),
      [propComponents],
    );

    const combinedContext = React.useMemo(
      () => ({ ...context, data }),
      [data, context],
    );

    return (
      <Virtuoso
        {...rest}
        {...(initialItemCount !== undefined ? { initialItemCount } : {})}
        topItemCount={stickyItemCount}
        ref={ref}
        useWindowScroll={useWindowScroll}
        customScrollParent={scrollParent ?? undefined}
        fixedItemHeight={fixedItemHeight}
        defaultItemHeight={estimatedItemHeight}
        className={className}
        data={data}
        context={combinedContext}
        components={components}
        itemContent={renderItemContent}
        computeItemKey={itemKey}
      />
    );
  },
);
