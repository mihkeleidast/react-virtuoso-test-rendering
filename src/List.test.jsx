import * as React from 'react';
import { render } from '@testing-library/react';

import { List } from './List';

describe('InfiniteList', () => {
  const data = [
    { id: '1', value: 'foo' },
    { id: '2', value: 'bar' },
    { id: '3', value: 'baz' },
  ];

  const itemKey = (_, { id }) => id;

  const Item = (props) => (
    <div {...props} data-testid={`item-${props.context?.data[props['data-item-index']]?.id}`} />
  );

  const itemContent = ({ item: { value } }) => <>{value}</>;

  it('correctly renders items', async () => {
    const { container, queryByTestId } = render(
      <List
        data={data}
        itemKey={itemKey}
        itemContent={itemContent}
        components={{ Item }}
        initialItemCount={data.length}
      />,
      { wrapper: ({ children }) => <div style={{ height: 300 }}>{children}</div> },
    );

    expect(container).toMatchSnapshot();

    // All items are rendered
    expect(queryByTestId('item-1')).not.toBeNull();
    expect(queryByTestId('item-2')).not.toBeNull();
    expect(queryByTestId('item-3')).not.toBeNull();
  });
});
