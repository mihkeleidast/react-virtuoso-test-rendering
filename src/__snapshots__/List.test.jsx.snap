// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfiniteList correctly renders items 1`] = `
<div>
  <div
    style="height: 300px;"
  >
    <div
      data-test-id="virtuoso-scroller"
      data-virtuoso-scroller="true"
      style="height: 100%; outline: none; overflow-y: auto; position: relative;"
      tabindex="0"
    >
      <div
        data-viewport-type="element"
        style="width: 100%; height: 100%; position: absolute; top: 0px;"
      >
        <div
          data-test-id="virtuoso-item-list"
          style="box-sizing: border-box; padding-top: 0px; padding-bottom: 0px; margin-top: 0px;"
        >
          <div
            context="[object Object]"
            data-index="0"
            data-item-index="0"
            data-known-size="0"
            data-testid="item-1"
          >
            foo
          </div>
          <div
            context="[object Object]"
            data-index="1"
            data-item-index="1"
            data-known-size="0"
            data-testid="item-2"
          >
            bar
          </div>
          <div
            context="[object Object]"
            data-index="2"
            data-item-index="2"
            data-known-size="0"
            data-testid="item-3"
          >
            baz
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
